client/allocwatcher/alloc_watcher.go

```diff
@@ -587,6 +587,14 @@ func (p *remotePrevAlloc) streamAllocDir(ctx context.Context, resp io.ReadCloser
 				p.prevAllocID, p.allocID, err)
 		}
 
+		escapes, err := escapingfs.PathEscapesAllocDir(dest, "", hdr.Name)
+		if err != nil {
+			return fmt.Errorf("error evaluating path: %w", err)
+		}
+		if escapes {
+			return fmt.Errorf("archive contains file that escapes alloc dir")
+		}
```

Comment:
How would a filesystem object escape the alloc dir if it's not a symlink?

Reply:
I was thinking in case a header `hdr` contains a relative path name or linkname

Reply:
> I was thinking in case a header `hdr` contains a relative path name or linkname

Ah, true. In practice the `hdr.Name` is always going to be relative because that's how the other end creates the archive. But this would add some protection against a compromised client that returns an evil archive instead.